File: architecture-skills/design-patterns/examples/interpreter/interpreter.js
// Interpreter pattern in JavaScript.
//
// A tiny arithmetic grammar: literals, variables, +, -, *. In JS the
// idiomatic form skips classes entirely: each grammar rule is a factory
// returning an evaluator function (env) => number, and composing
// factories builds the expression tree.

const literal = (value) => () => value;
const variable = (name) => (env) => env[name];
const add = (left, right) => (env) => left(env) + right(env);
const sub = (left, right) => (env) => left(env) - right(env);
const mul = (left, right) => (env) => left(env) * right(env);

// (x + y) * 2 - z
const expression = sub(mul(add(variable("x"), variable("y")), literal(2)), variable("z"));

const contexts = [
  { x: 6, y: 2, z: 3 },
  { x: 1, y: 1, z: 10 },
];

for (const context of contexts) {
  console.log(`(x + y) * 2 - z with ${JSON.stringify(context)} = ${expression(context)}`);
}

// Class-based nodes (the literal GoF form) are only worth it when the
// tree must be inspected, transformed, or serialized — closures are
// opaque. A minimal node form for that case:
const node = {
  interpret: (env) => env.x + 2,
  toString: () => "x + 2",
};
console.log(`${node} with x=40 = ${node.interpret({ x: 40 })}`);

File: architecture-skills/design-patterns/examples/interpreter/interpreter.py
"""Interpreter pattern in Python.

A tiny arithmetic grammar: literals, variables, +, -, *. Each grammar
rule is a node class with interpret(context); a sentence is an object
tree evaluated against a dict of variable bindings.

Note: when trees never need to be inspected or serialized, a plain
lambda (lambda env: env["x"] + 2) is often the simpler Python answer.
"""

from __future__ import annotations

from dataclasses import dataclass

Context = dict[str, int]


@dataclass(frozen=True)
class Literal:
    value: int

    def interpret(self, context: Context) -> int:
        return self.value


@dataclass(frozen=True)
class Variable:
    name: str

    def interpret(self, context: Context) -> int:
        return context[self.name]


@dataclass(frozen=True)
class Add:
    left: Literal | Variable | Add | Sub | Mul
    right: Literal | Variable | Add | Sub | Mul

    def interpret(self, context: Context) -> int:
        return self.left.interpret(context) + self.right.interpret(context)


@dataclass(frozen=True)
class Sub:
    left: Literal | Variable | Add | Sub | Mul
    right: Literal | Variable | Add | Sub | Mul

    def interpret(self, context: Context) -> int:
        return self.left.interpret(context) - self.right.interpret(context)


@dataclass(frozen=True)
class Mul:
    left: Literal | Variable | Add | Sub | Mul
    right: Literal | Variable | Add | Sub | Mul

    def interpret(self, context: Context) -> int:
        return self.left.interpret(context) * self.right.interpret(context)


if __name__ == "__main__":
    # (x + y) * 2 - z
    expression = Sub(Mul(Add(Variable("x"), Variable("y")), Literal(2)), Variable("z"))

    for context in ({"x": 6, "y": 2, "z": 3}, {"x": 1, "y": 1, "z": 10}):
        print(f"(x + y) * 2 - z with {context} = {expression.interpret(context)}")
